### PR TITLE
[OV JS] Support Core.setProperty() & Core.getProperty() methods in JS API

### PR DESCRIPTION
**Pull request by vishniakov-nikolai**
### Details:
 - Implement Core.setProperty()
 - Implement Core.getProperty()

### Tickets:
 - 134824

### CODE REVIEW

File: src/bindings/js/node/tests/core.test.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  core.setProperty({ 'CACHE_DIR': tmpDir });
+
+  const cacheDir = core.getProperty('CACHE_DIR');
+
+  assert.equal(cacheDir, tmpDir);
+});
+
+it('Core.setProperty(\'CPU\')', () => {    
+  const tmpDir = '/home/nvishnya/tmp';
```
Comment (almilosz) [inline src/bindings/js/node/tests/core.test.js:22]: Please remove this hardcoded path

File: src/bindings/js/node/tests/core.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+});
+
+it('Core.getProperty(\'CPU\', \'SUPPORTED_PROPERTIES\') is Array', () => {    
+  const supportedPropertiesArray = core.getProperty('CPU', 'SUPPORTED_PROPERTIES');
+
+  assert.ok(Array.isArray(supportedPropertiesArray));
+});
+
+it('Core.setProperty(\'CPU\', { \'NUM_STREAMS\': 5 })', () => {   
+  const streams = 5;
```
Comment (almilosz) [inline src/bindings/js/node/tests/core.test.js:38]: Why number of streams is constant?

File: src/bindings/js/node/include/core_set_property_args.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+struct CoreSetPropertyArgs {
+    std::string device_name;
+    ov::AnyMap parameters;
+
+    CoreSetPropertyArgs() {}
+    CoreSetPropertyArgs(const Napi::CallbackInfo& info) {
+        const size_t args_length = info.Length();
+
+        if (!is_valid_input(args_length, info))
```
Comment (almilosz) [inline src/bindings/js/node/include/core_set_property_args.hpp:18]: I think the exception can be thrown from the `is_valid_input` method.  If-statement with the negation of the condition is less clear and unnecessary.

File: src/bindings/js/node/include/core_set_property_args.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            throw std::runtime_error("Invalid arguments of set_property function");
+
+        if (args_length > 1) device_name = info[0].ToString();
+        
+        Napi::Object parameters = info[args_length > 1 ? 1 : 0].ToObject();
```
Comment (almilosz) [inline src/bindings/js/node/include/core_set_property_args.hpp:23]: Does this condition apply to a situation where no `device_name` is specified?

File: src/bindings/js/node/src/core_wrap.cpp [hunk 5]
```diff
@@ -230,3 +233,46 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
 
     return js_devices;
 }
+
+Napi::Value CoreWrap::set_property(const Napi::CallbackInfo& info) {
+    try {
+        CoreSetPropertyArgs* args;
+        args = new CoreSetPropertyArgs(info);
+
+        if (args->device_name.empty()) {
```
Comment (almilosz) [inline src/bindings/js/node/src/core_wrap.cpp:242]: Here is also a check whether `device_name` is specified

File: src/bindings/js/node/include/core_set_property_args.hpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            auto property_name = static_cast<Napi::Value>(keys[i]).ToString().Utf8Value();
+            
+
+            ov::Any any_value = js_to_any(info, parameters.Get(property_name));
+
+            this->parameters.insert(std::make_pair(property_name, any_value));
+        }
+    }
+
+    bool is_valid_input(size_t args_length, const Napi::CallbackInfo& info) {
```
Comment (almilosz) [inline src/bindings/js/node/include/core_set_property_args.hpp:36]: What is this function validating? Length of input? If `params/device_name` were passed?

**Review by almilosz — COMMENTED** [review]
Please take a look at whether logical conditions can be simplified

File: src/bindings/js/node/src/core_wrap.cpp (hunk 5 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/src/core_wrap.cpp:242]: Any proposals?

File: src/bindings/js/node/include/core_set_property_args.hpp (hunk 6 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/include/core_set_property_args.hpp:36]: Arguments passed from js part)

File: src/bindings/js/node/include/core_set_property_args.hpp (hunk 3 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/include/core_set_property_args.hpp:18]: Then, it has call earlier, from original method for example

File: src/bindings/js/node/tests/core.test.js (hunk 1 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/tests/core.test.js]: Done

File: src/bindings/js/node/src/core_wrap.cpp [hunk 7]
```diff
@@ -230,3 +233,47 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
 
     return js_devices;
 }
+
+Napi::Value CoreWrap::set_property(const Napi::CallbackInfo& info) {
+    try {
+        CoreSetPropertyArgs* args;
+        args = new CoreSetPropertyArgs(info);
```
Comment (praasz) [inline src/bindings/js/node/src/core_wrap.cpp:240]: Why using `new` is required?

File: src/bindings/js/node/include/core_set_property_args.hpp [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <napi.h>
+
+#include "node/include/helper.hpp"
+#include "openvino/runtime/core.hpp"
+
+/**
+ * @brief This struct retrieves data from Napi::CallbackInfo.
+ */
+struct CoreSetPropertyArgs {
```
Comment (praasz) [inline src/bindings/js/node/include/core_set_property_args.hpp:11]: Maybe I don't see purpose for this struct but from usage it looks like could be replaced by set of functions like
- validate
- get_paramters
- get_device_name

or maybe function which tries get some property from `CallbackInfo` or throw.
 - try_get_paramters
 - try_get_device_name

File: src/bindings/js/node/include/core_set_property_args.hpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
+#pragma once
```
Comment (praasz) [inline src/bindings/js/node/include/core_set_property_args.hpp:1]: Add Copyright comment.

**Review by praasz — CHANGES_REQUESTED** [review]

File: src/bindings/js/node/src/core_wrap.cpp (hunk 7 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/src/core_wrap.cpp]: Not required. Replaced.

File: src/bindings/js/node/include/core_set_property_args.hpp (hunk 8 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/include/core_set_property_args.hpp]: Main purpose was to improve readability and do not full core_wrap.cpp by extra helpers. But I agree, structure in separate file looks redundant. 
Refactored.

File: src/bindings/js/node/src/helper.cpp [hunk 10]
```diff
@@ -318,3 +318,210 @@ ov::Tensor cast_to_tensor(const Napi::TypedArray& typed_array,
...
+template <typename T>
+Napi::Array cpp_vector_to_js_array(const Napi::CallbackInfo& info, const std::vector<T>& vec) {
+    Napi::Array array = Napi::Array::New(info.Env(), vec.size());
+
+    uint32_t i = 0;
+    for (auto& property : vec) {
+        auto any = ov::Any(property);
+
+        array[i++] = any_to_js(info, any);
+    }
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/bindings/js/node/src/helper.cpp]: Yes, I've forgot about capture. Will ` [&info](const auto& property){` work? if not don't change it is detail.

File: src/bindings/js/node/src/core_wrap.cpp [hunk 11]
```diff
@@ -11,6 +11,44 @@
 #include "node/include/model_wrap.hpp"
 #include "node/include/read_model_args.hpp"
 
+void validate_set_property_args(const Napi::CallbackInfo& info) {
+    const size_t args_length = info.Length();
+    const bool is_device_specified = info[0].IsString();
+    const bool has_params_obj = info[is_device_specified ? 1 : 0];
+
+    if (!has_params_obj)
+        throw std::runtime_error("Properties parameter must be an object");
```
Comment (almilosz) [inline src/bindings/js/node/src/core_wrap.cpp:20]: `OPENVINO_THROW`

File: src/bindings/js/node/src/helper.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (vishniakov-nikolai) [reply src/bindings/js/node/src/helper.cpp]: With `[&info](const auto& property){` it's also doesn't work.
Error message: 
```
/home/n/openvino/src/bindings/js/node/src/helper.cpp:411:46:   required from here
/usr/include/c++/9/bits/stl_algo.h:4343:12: error: no match for ‘operator=’ (operand types are ‘std::pair<Napi::Value, Napi::Object::PropertyLValue<Napi::Value> >’ and ‘Napi::Value’)
 4343 |  *__result = __unary_op(*__first);
      |  ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/9/utility:70,
                 from /usr/include/c++/9/tuple:38,
                 from /usr/include/c++/9/functional:54,
                 from /home/n/openvino/build/_deps/node-addon-api-src/napi.h:14,
                 from /home/n/openvino/src/bindings/js/node/../node/include/helper.hpp:5,
                 from /home/n/openvino/src/bindings/js/node/src/helper.cpp:4:
/usr/include/c++/9/bits/stl_pair.h:378:7: note: candidate: ‘std::pair<_T1, _T2>& std::pair<_T1, _T2>::operator=(typename std::conditional<std::__and_<std::is_copy_assignable<_T1>, std::is_copy_assignable<_T2> >::value, const std::pair<_T1, _T2>&, const std::__nonesuch_no_braces&>::type) [with _T1 = Napi::Value; _T2 = Napi::Object::PropertyLValue<Napi::Value>; typename std::conditional<std::__and_<std::is_copy_assignable<_T1>, std::is_copy_assignable<_T2> >::value, const std::pair<_T1, _T2>&, const std::__nonesuch_no_braces&>::type = const std::pair<Napi::Value, Napi::Object::PropertyLValue<Napi::Value> >&]’
  378 |       operator=(typename conditional<
      |       ^~~~~~~~
/usr/include/c++/9/bits/stl_pair.h:381:51: note:   no known conversion for argument 1 from ‘Napi::Value’ to ‘std::conditional<true, const std::pair<Napi::Value, Napi::Object::PropertyLValue<Napi::Value> >&, const std::__nonesuch_no_braces&>::type’ {aka ‘const std::pair<Napi::Value, Napi::Object::PropertyLValue<Napi::Value> >&’}
  378 |       operator=(typename conditional<
      |                 ~~~~~~~~~~~~~~~~~~~~~
  379 |   __and_<is_copy_assignable<_T1>,
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  380 |          is_copy_assignable<_T2>>::value,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  381 |   const pair&, const __nonesuch_no_braces&>::type __p)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
```

File: src/bindings/js/node/src/helper.cpp [hunk 12]
```diff
@@ -318,3 +318,221 @@ ov::Tensor cast_to_tensor(const Napi::TypedArray& typed_array,
...
+
+    return info.Env().Undefined();
+}
+
+ov::Any js_to_any(const Napi::CallbackInfo& info, Napi::Value value) {
+    if (value.IsString()) {
+        return ov::Any(value.ToString().Utf8Value());
+    } else if (value.IsBigInt()) {
+        Napi::BigInt big_value = value.As<Napi::BigInt>();
+        bool loses;
```
[the remark this answers is not in the document]
Comment (vishniakov-nikolai) [reply src/bindings/js/node/src/helper.cpp]: Done

File: src/bindings/js/node/src/helper.cpp [hunk 13]
```diff
@@ -318,3 +318,225 @@ ov::Tensor cast_to_tensor(const Napi::TypedArray& typed_array,
...
+
+    return info.Env().Undefined();
+}
+
+ov::Any js_to_any(const Napi::CallbackInfo& info, Napi::Value value) {
+    if (value.IsString()) {
+        return ov::Any(value.ToString().Utf8Value());
+    } else if (value.IsBigInt()) {
+        Napi::BigInt big_value = value.As<Napi::BigInt>();
+        bool isLossless;
```
Comment (almilosz) [inline src/bindings/js/node/src/helper.cpp:509]:
```suggestion
        bool is_lossless
```

**Review by praasz — APPROVED** [review]

**Pull request merged**